src/vmx.rs

```diff
@@ -54,11 +54,18 @@ impl Vmx {
             rflags
         };
 
-        // FIXME: this leaks the page on error
-        error::check_vm_insruction(rflags, "Failed to enable vmx".into())?;
-        Ok(Vmx {
-            vmxon_region: vmxon_region,
-        })
+        match error::check_vm_insruction(rflags, "Failed to enable vmx".into()) {
```

Comment:
I think I would rather this be written as `error::check_vm_insruction(...).map_err(|e| { ... deallocate ... })?`

Reply:
I agree that this would be much cleaner, but borrowck doesn't like that `vmxon_region` is moved in the closure and assigned to the `vmxon_region` member. When using the `match` it is easier for it to determine that it is unable to get to the `Ok` block if we deallocate the region.

Reply:
Ah, fair port. Maybe something like this then (just getting rid of the explicit unit and unused `res`):

```suggestion
        match error::check_vm_insruction(rflags, "Failed to enable vmx".into()) {
            Ok(_) => Ok(Vmx {
                vmxon_region: vmxon_region,
            }),
            Err(e) => {
                alloc.deallocate_frame(vmxon_region).unwrap_or_else(|_| {
                    info!("Failed to deallocate vmxon region");
                });
                Err(e)
            }
        }
```

At least until we have something set up to preserve context.

Reply:
:+1: updated.